### PR TITLE
[pathval] Add types for pathval

### PR DESCRIPTION
**Pull request by 9999years**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/pathval/tslint.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
```
Comment (sandersn) [inline types/pathval/tslint.json]: New packages should not override any lint rules. They should only extend `dtslint/dt.json`.

File: types/pathval/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        hasProperty(obj: object | undefined | null, name: Property): boolean;
+        getPathInfo(obj: object, path: string): PathInfo;
+        getPathValue(obj: object, path: string): object | undefined;
+        setPathValue(obj: object, path: string, val: any): object;
+    }
+}
+
+declare const pathval: PathVal.PathValStatic;
+
+declare module "pathval" {
```
Comment (sandersn) [inline types/pathval/index.d.ts]: 1. Please use the modern UMD syntax to create the global object name: `export as namespace pathval`. 
2. This means you can delete `declare const pathval`.
3. Also delete `declare namespace PathVal` and dedent its contents. Prefix the types with `export`. 
4. Also delete `interface PathValStatic`. Dedent its contents and prefix them with `export function`.

**Review by sandersn — CHANGES_REQUESTED** [review]
Needs a few structural changes, but looks good otherwise.

**Comment by 9999years** [comment]
OK, made the requested changes. Should `export as namespace pathval` be at the start of the file rather than the end?

PTAL 🙂

**Comment by sandersn** [comment]
I don't have a preference. I usually put it at the end.

**Review by sandersn — APPROVED** [review]

**Pull request merged**